server/src/main/java/io/unitycatalog/server/service/credential/CredentialOperations.java

```diff
@@ -0,0 +1,108 @@
+
+import java.net.URI;
+import java.util.List;
+import java.util.Set;
+
+import static io.unitycatalog.server.utils.Constants.URI_SCHEME_ABFS;
+import static io.unitycatalog.server.utils.Constants.URI_SCHEME_ABFSS;
+import static io.unitycatalog.server.utils.Constants.URI_SCHEME_GS;
+import static io.unitycatalog.server.utils.Constants.URI_SCHEME_S3;
+
+public class CredentialOperations {
+
+  private final AwsCredentialVendor awsCredentialVendor;
+  private final AzureCredentialVendor azureCredentialVendor;
+  private final GcpCredentialVendor gcpCredentialVendor;
+
+  public CredentialOperations() {
+    this.awsCredentialVendor = new AwsCredentialVendor();
+    this.azureCredentialVendor = new AzureCredentialVendor();
+    this.gcpCredentialVendor = new GcpCredentialVendor();
+  }
+
+  public GenerateTemporaryTableCredentialResponse vendCredentialForTable(TableInfo table) {
+    String tableStorageLocation = table.getStorageLocation();
+    if (tableStorageLocation == null || tableStorageLocation.isEmpty()) {
+      throw new BaseException(ErrorCode.FAILED_PRECONDITION, "Table storage location not found.");
+    }
+
+    URI storageLocationUri = URI.create(tableStorageLocation);
+
+    // FIXME!! Update privileges when access controls are implemented
+    CredentialContext credentialContext = CredentialContext.builder().storageScheme(storageLocationUri.getScheme()).storageBase(storageLocationUri.getPath())
+      .privileges(Set.of(CredentialContext.Privilege.SELECT, CredentialContext.Privilege.UPDATE)).locations(List.of(table.getStorageLocation())).build();
+
+    return vendCredential(credentialContext).toTableCredentialResponse();
+  }
+
+  public GenerateTemporaryVolumeCredentialResponse vendCredentialForVolume(VolumeInfo volume) {
+    String volumePath = volume.getStorageLocation();
+    if (volumePath == null || volumePath.isEmpty()) {
+      throw new BaseException(ErrorCode.FAILED_PRECONDITION, "Volume storage location not found.");
+    }
+
+    URI storageLocationUri = URI.create(volumePath);
+
+    // FIXME!! Update privileges when access controls are implemented
+    CredentialContext credentialContext = CredentialContext.builder().storageScheme(storageLocationUri.getScheme()).storageBase(storageLocationUri.getPath())
+      .privileges(Set.of(CredentialContext.Privilege.SELECT, CredentialContext.Privilege.UPDATE)).locations(List.of(volume.getStorageLocation())).build();
+
+    return vendCredential(credentialContext).toVolumeCredentialResponse();
+  }
+
+  public CredentialResponse vendCredential(CredentialContext context) {
+    CredentialResponse.CredentialResponseBuilder builder = new CredentialResponse.CredentialResponseBuilder();
+
+    switch (context.storageScheme) {
+      case URI_SCHEME_ABFS, URI_SCHEME_ABFSS -> {
+        AzureCredential azureCredential = vendAzureCredential(context);
+        builder.azureUserDelegationSas(new AzureUserDelegationSAS().sasToken(azureCredential.getSasToken()))
```

Comment:
nit: why not move the builder into each vend method (like `vendAzureCredential()` since those methods aren't really doing much themselves.

Reply:
yeah, thought about this, only thing is there are other places (specifically on the Iceberg REST side that don't need to use the Unity API specific models, but they could I guess)